### PR TITLE
Deprecate public Controller properties

### PR DESCRIPTION
**Pull request by robertpustulka**
Refs #10946

I'm afraid it's hard to change the visibility of `$request` and `$response` properties now, so I kept them public but added deprecation annotations to them.

`$name`, `$plugin`, `$autoRender` properties are now protected and accessed via magic methods for BC.

It was also hard to refactor the code to use request controller name and plugin params as those not always share the same state, at least in tests. So these are still controller properties. We can get back to this in 4.0 though.

I added deprecation annotations to `$components` and `$helpers` as well because those are deprecated since the 3.0 I believe.

After this is reviewed and merged, I'll update the core to use new methods.

### CODE REVIEW

File: src/Controller/Controller.php
Comment (chinpei215) [inline src/Controller/Controller.php]: I think this change is meaningless. Because when I touch this property, I am in the action of the controller usually. So I can continue to use this property without any warnings even if I am using some great IDE. Probably, it's the time to allow us to use `private` modifiers. To be honest, I think clean APIs wouldn't prefer setters/getters for public/protected properties or getters for private mutable objects.

**Comment by inoas** [comment]
I'd either have also added $request and $response to the magic getters/setters or (preferably) not used magic getters/setters at all but just mark the properties with a deprecation warning that their visibility will change from public to private in 4.0 and that you should use the appropriate getter/setter methods.

**Comment by robertpustulka** [comment]
> I'd either have also added $request and $response to the magic getters/setters

I tried, but many of the tests just failed. I'm afraid it's not that easy to do this.


> Because when I touch this property, I am in the action of the controller usually.

For me that is a reason this property shouldn't be public.

File: src/Controller/Controller.php
Comment (inoas) [inline src/Controller/Controller.php]: This is one of the places where I would strongly prefer `request()` for getting as that is the 99% case and everything else is annoying verbosity. See https://github.com/cakephp/cakephp/issues/10930.

I am not even sure if the request object should be changed? Should it? Or should it merely be cloned via `withFoo()`?

File: src/Controller/Controller.php
Comment (ADmad) [inline src/Controller/Controller.php]: This change is not BC and actually not meaningless. There could be some doing `$this->getController()->autoRender = false` in a component.

**Comment by ADmad** [comment]
Changing public properties to protected in not backwards compatible, please revert it. I showed an example above how it can create problems.

**Comment by robertpustulka** [comment]
@ADmad There are magic `__get` and `__set` updated to handle this. Please look at the diff :)

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: @ionas I prefer the consistency. All getters here have a prefix, so this one should also have it. BTW `setRequest` has been around for the while and is used to set the state of the controller.

**Comment by ADmad** [comment]
Sorry missed that :)

File: src/Controller/Controller.php
Comment (ADmad) [inline src/Controller/Controller.php]: If this method has `enable` prefix then it should only enable auto rendering. Currently doing `enableAutoRender(false)` would disable auto rendering which is not good for readability.

**Review by ADmad — CHANGES_REQUESTED** [review]

**Comment by chinpei215** [comment]
What do you think about benefits of setters/getters? In my opinion, setters ensure to set unexpected values to the property. But if it is a `protected` property, I can write `$this->autoRender = 0;` in my controller for example. So this kind of setters/getters don't have any advantages actually.

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: I was following the examples in a query builder, like `enableAutoFields` and `enableHydration` which accept `false`.

**Comment by robertpustulka** [comment]
@chinpei215 Right, I can add a typecheck to the `enableAutoRender` since we're not on PHP 7 yet.

**Comment by chinpei215** [comment]
Even if we are on PHP7, properties will be weakly typed. Are you planning to change the modifier to `private` in 4.0? Then I am ok. But if you are not planning that, I am not in favor of this change, as people would continue to write `$this->autoRender = true;`.

File: src/Controller/Controller.php
Comment (ADmad) [inline src/Controller/Controller.php]: So either the method should be name `setAutoRender()` or two separate methods `enableAutoRender()` and `disableAutoRender()` should be added which don't take any argument.

**Comment by robertpustulka** [comment]
@chinpei215 I can't recall if there has been any discussion or decision made on `private` properties. But I also think that some of the properties could become private.

**Comment by chinpei215** [comment]
OK. Thank you for your reply :)

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: @ADmad are you suggesting that we should do the same in a query builder? That obviously makes more sense then calling `enableSomething(false)`

Comment (markstory) [inline src/Controller/Controller.php]: I've found `enableThing(false)` to be clear enough, but I might be an outlier there.

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: Because responses and requests are immutable I'm wondering if we should start thinking of ways to make working with responses less clunky. Using setResponse() is going to get annoying.

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: Subclasses could call `$this->enable/disableAutoRender()` in their initialize methods.

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: We could but I've seen some tests that set the plugin property. I'm not sure how this would work with immutable requests but this could be a matter of refactoring tests.

File: src/Controller/Controller.php
Comment (robertpustulka) [inline src/Controller/Controller.php]: Are there many cases where a request is modified in a controller scope?

File: src/Controller/Controller.php
Comment (ADmad) [inline src/Controller/Controller.php]: For me personally even `something(false)` is good enough :)

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: They could. But making just _this_ property private would introduce inconsistencies with other parts of the framework and break compatibility.

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: We can start by refactoring our tests.

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: I'm sure there are. People do things like mutate request data quite often.

Comment (robertpustulka) [inline src/Controller/Controller.php]: Right, I forgot about data. 

In controller methods it will be still possible to do `$this->request = $this->request->withSomething()`

**Comment by saeideng** [comment]
I think 
1)we need change scope of `request` and `response` to protected
2) un-deprecate   accessing `this->request` and `$this->response`
3)set those by `__set()` and validate if received object is `serverRequest` and `response` Object or no
4)run a error if received object is invalid 
5)get by `__get()` 

so we able to use 
`$this->request = $this->request->withSomething()`
`$this->response  = $this->response->withSomething()`
safety

**Comment by chinpei215** [comment]
No. When you assign a value to a protected property from a derived class, `__set()` will not be called. So people can write `$this->request = 1;` unsafely. Please read the earlier discussion about `private` modifier.

**Comment by saeideng** [comment]
what happen if we change `public $request` to `protected $_request` or `protected $_other_request_property`

**Comment by saeideng** [comment]
i means
```
     public function setRequest(ServerRequest $request)
     {
         $this->request = $request;
         $this->plugin = $request->getParam('plugin') ?: null;
 
         if ($request->getParam('pass')) {
             $this->passedArgs = $request->getParam('pass');
         }

        return $this;
    }
```
```
         $deprecated = [
            'request' => 'setRequest',
            'response' => 'setResponse'
        ];
        if (isset($deprecated[$name])) {
            $method = $deprecated[$name];

            return $this->{$method}();
        }
```

**Comment by saeideng** [comment]
you can not pass `1` to `public function setRequest(ServerRequest $request)`
I misunderstood ?

**Comment by chinpei215** [comment]
> I misunderstood ?

Yes. First, people can access the protected property directly: `$this->_request = 1;`.
Second, if you declare `public $request = 1;` in a derived class, the property is evaluated as `1` in any methods of the parent class **if `private $request` is not declared** . So, for example, `$this->referer()` will cause fatal error because it calls `$this->request->referer()` internally.

**Comment by saeideng** [comment]
I forgot we are in subcluss of cor controller :smiley:

**Comment by robertpustulka** [comment]
I've added the `disableAutoRender()` method and marked the property deprecated. It may convince users to using new  methods in their actions instead.
I've also removed get/setPlugin methods, deprecated the property in favor of the requets param.

I'm going to remove the deprecated properties and method calls in another PR once this is merged.

**Comment by robertpustulka** [comment]
Is there a need for deprecation warnings for the properties in __get and __set? I mean something like with the deprecated view builder properties.

**Comment by cleptric** [comment]
@robertpustulka I would add them after 3.5 shipped.

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: This could use `?:` now I think.

File: tests/TestCase/Controller/ControllerTest.php
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: Looks like a mistake.

**Review by ADmad — APPROVED** [review]

**Comment by markstory** [comment]
I am going to hold off merging this until 3.6 development starts. 3.5 is already a big release.

**Comment by dereuromark** [comment]
Looks like this is now pretty much good to go.

File: src/Controller/Controller.php
Comment (markstory) [reply src/Controller/Controller.php]: @inoas You wouldn't need to use the methods *inside* the controller actions. In that context you could use the protected properties.

File: src/Controller/Controller.php
Comment (markstory) [inline src/Controller/Controller.php]: These numbers should be `3.6.0` now.

**Comment by robertpustulka** [comment]
I've updated the numbers, added `@since` tags and fixed some tests to use getters and setters.

File: src/Controller/Controller.php
Comment (dereuromark) [inline src/Controller/Controller.php]: Will this be a BC issue from components setting this on the controller?

Comment (robertpustulka) [reply src/Controller/Controller.php]: With this it should be ok:
https://github.com/robertpustulka/cakephp/blob/e2d894e65b7731b971a8ded78d42971bb32b7c27/src/Controller/Controller.php#L403-L407

File: tests/TestCase/Controller/ControllerTest.php
Comment (saeideng) [inline tests/TestCase/Controller/ControllerTest.php]: `$message` not used 
see my recent PR

Comment (robertpustulka) [reply tests/TestCase/Controller/ControllerTest.php]: I'll wait with updating tests until #11231 is merged if that's what you mean.

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: yes 
I'll update it tonight

**Comment by robertpustulka** [comment]
The branch is up to date with `3.next`.

I added `plugin` getter and setter to the controller. Some tests seem to expect that the state of controller plugin and request plugin parameters may differ and I didn't want to break that behavior by delegating controller plugin property to the request entirely.

This could be refactored in 4.x so the `name` and `plugin` parameters would be accessed only via request.

File: tests/TestCase/Controller/ControllerTest.php
Comment (saeideng) [inline tests/TestCase/Controller/ControllerTest.php]: if structure of this and `deprecatedViewPropertyProvider` is same
you can reduce test by adding content of this to `deprecatedViewPropertyProvider` 
so you can remove  bellow test IMO
```php
testDeprecatedControllerPropertyGetterMessage
testDeprecatedControllerPropertySetterMessage
testDeprecatedControllerProperty
```
also `expectedExceptionMessageRegExp /^Controller::\$\w+ is deprecated(.*)/` is same for 9 deprecated methods and this RegEx just check first of message (until `'Controller::$%s is deprecated.` )not all

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: we no need to check all `'Controller::$%s is deprecated. Use $this->%s instead.'` message IMO

Comment (robertpustulka) [reply tests/TestCase/Controller/ControllerTest.php]: Deprecated view properties point to the `$controller->viewBuilder()->method()` so this is a bit different.

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: OK :+1:  Your are right 
I forget the `->viewBuilder()`

**Comment by robertpustulka** [comment]
Can it be merged? I'd like to do the similar with `View` and `Cell` classes.

File: tests/TestCase/Controller/ControllerTest.php
Comment (saeideng) [inline tests/TestCase/Controller/ControllerTest.php]: now
you should use `withErrorReporting()` for wrapping   `error_reporting()`
and `deprecated()` for testing deprecated methods

in here you must use `deprecated()`

File: src/Controller/Controller.php
Comment (ADmad) [inline src/Controller/Controller.php]: I don't think we should `getPlugin()`, `setPlugin()`. Just make people use `$this->request->getParam('plugin')` instead.

Comment (robertpustulka) [reply src/Controller/Controller.php]: I tried this but some tests were failing. It looked to me that there were some cases where different state od request plugin param and controller plugin name was expected.

Comment (ADmad) [reply src/Controller/Controller.php]: Okay, that's sad :disappointed:.

File: tests/TestCase/Controller/ControllerTest.php
Comment (robertpustulka) [reply tests/TestCase/Controller/ControllerTest.php]: @saeideng tests have been updated

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: nice :+1:

**Comment by robertpustulka** [comment]
Can we merge this? I had to resolve conflicts like 3rd or 4th time already...

File: tests/TestCase/Controller/ControllerTest.php
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: The default error level is `E_ALL` why do you need to increase it here?

Comment (robertpustulka) [reply tests/TestCase/Controller/ControllerTest.php]: Must be some out of date code. I'll fix that.

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: default yes 
but we can change default in future if needed 
and in this case test must return  error/warning , so `E_ALL` is true IMO

Comment (robertpustulka) [reply tests/TestCase/Controller/ControllerTest.php]: @markstory I think that @saeideng has a point here. What is your opinion on this?

Comment (markstory) [reply tests/TestCase/Controller/ControllerTest.php]: I would prefer to keep our test suite at `E_ALL` and disable deprecation warnings as needed via `deprecated()`.

Comment (saeideng) [reply tests/TestCase/Controller/ControllerTest.php]: so we no need to assert deprecation warnings in tests like other place

**Pull request merged**
